### ISSUE TITLE
shared libraries building support

### ISSUE THREAD

**Issue by rjktcby** [KhronosGroup/KTX#18]
Couldn't regenerate cmake build system on my debian 8 quickly (some python import errors), so firing up an issue here.

On Linux it's important to be able to link against a shared library, so adding support to build all target libs shared would be great. Seems like it shouldn't be a big deal from what i've read in [gyp docs](https://gyp.gsrc.io/docs/UserDocumentation.md):
`'type': This should almost always be set to ‘<(library)’, which allows the user to define at gyp time whether libraries are to be built static or shared`
It's also important to add libGL to dependencies (at least for libktx_gl target) otherwise shared library build will fail

Think I can produce a patch myself if i get some tip on how to make gyp work on my machine. Currently getting this:

    make cmake
    gyp -f cmake -DOS=linux --generator-output=build/cmake/linux/ -G output_dir=. --depth=. ktxtests.gyp ktxtools.gyp
    Traceback (most recent call last):
      File "/usr/bin/gyp", line 9, in <module>
        load_entry_point('gyp==0.1', 'console_scripts', 'gyp')()
      File "/usr/lib/python2.7/dist-packages/gyp/__init__.py", line 547, in script_main
        return main(sys.argv[1:])
      File "/usr/lib/python2.7/dist-packages/gyp/__init__.py", line 540, in main
        return gyp_main(args)
      File "/usr/lib/python2.7/dist-packages/gyp/__init__.py", line 516, in gyp_main
        options.circular_check)
      File "/usr/lib/python2.7/dist-packages/gyp/__init__.py", line 91, in Load
        generator = __import__(generator_name, globals(), locals(), generator_name)
    ImportError: No module named cmake
    GNUmakefile:130: recipe for target 'build/cmake/linux/.ktx-stamp' failed
    make: *** [build/cmake/linux/.ktx-stamp] Error 1

**Comment by msc-** [comment]
Are you using a distribution of gyp that is included with Debian? After I run `sudo ./setup.py install` in my clone of GYP on Ubuntu, gyp is installed in /`usr/local/lib/python2.7/dist-packages`.

The cmake generator has been included in GYP for the entire 2+ years I have been using it so I don't know why it should be missing from the Debian distribution, if that is what you are using. Check your `/usr/lib/python2.7/dist-packages/gyp/generator` for `cmake.{py,pyc}`.

I suggest you clone the GYP source and install it yourself. You can find my fork [here](https://github.com/msc-/gyp). Checkout branch `remaster`. Or you can find the upstream at Google [here](https://chromium.googlesource.com/external/gyp).

I did not add a dependency on libGL because applications using the writeKTX functions, e.g. toktx, do not need it. So please only add the dependency if building a libktx shared library. You can use a `target_conditions` block like this inside the libktx target.

```
'target_conditions': [
  ['_type == "shared_library"', {
     'includes': [ '../gyp_include/libgl.gypi' ],
     'dependencies':  [ 'libgl' ],
  }],
]

**Comment by rjktcby** [comment]
Actually i've already solved the issue with cmake gyp generator. Yes, it is not present in debian gyp package and i've built gyp from source so now i have it. Can't make your `target_conditions` work, getting errors.

Here's my diff:

    diff --git a/lib/libktx.gypi b/lib/libktx.gypi
    index a3bf9a3..eafabf5 100644
    --- a/lib/libktx.gypi
    +++ b/lib/libktx.gypi
    @@ -43,7 +43,7 @@
             # Because these must be specified in two places.
             'defines': [ 'KTX_OPENGL=1' ],
           },
    -      'type': 'static_library',
    +      'type': '<(library)',
           'defines': [ '<@(defines)' ],
           'direct_dependent_settings': {
              'defines': [ '<@(defines)' ],
    @@ -51,6 +51,12 @@
           },
           'sources': [ '<@(sources)' ],
           'include_dirs': [ '<@(include_dirs)' ],
    +      'target_conditions': [
    +        ['_type == "shared_library"', {
    +          'includes': [ '../gyp_include/libgl.gypi' ],
    +          'dependencies': ['libgl'],
    +        }]
    +      ],
         }, # libktx.gl target
         {
           'target_name': 'libktx.es1',

Here's gyp output

    gyp -f cmake -Dlibrary=shared_library -DOS=linux --generator-output=build/cmake/linux/ -G output_dir=. --depth=. ktxtests.gyp ktxtools.gyp
    Traceback (most recent call last):
      File "/usr/local/bin/gyp", line 9, in <module>
        load_entry_point('gyp==0.1', 'console_scripts', 'gyp')()
      File "/usr/local/lib/python2.7/dist-packages/gyp-0.1-py2.7.egg/gyp/__init__.py", line 545, in script_main
        return main(sys.argv[1:])
      File "/usr/local/lib/python2.7/dist-packages/gyp-0.1-py2.7.egg/gyp/__init__.py", line 538, in main
        return gyp_main(args)
      File "/usr/local/lib/python2.7/dist-packages/gyp-0.1-py2.7.egg/gyp/__init__.py", line 523, in gyp_main
        generator.GenerateOutput(flat_list, targets, data, params)
      File "/usr/local/lib/python2.7/dist-packages/gyp-0.1-py2.7.egg/gyp/generator/cmake.py", line 1241, in GenerateOutput
        pool.map(CallGenerateOutputForConfig, arglists)
      File "/usr/lib/python2.7/multiprocessing/pool.py", line 251, in map
        return self.map_async(func, iterable, chunksize).get()
      File "/usr/lib/python2.7/multiprocessing/pool.py", line 558, in get
        raise self._value
    KeyError: 'libgl'

If i comment out the `'dependencies'` line, gyp succeeds, but libGL linkage line is not generated in CMakeLists.txt.
I'm new to gyp, don't know yet how to effectively debug it, python stacktraces are not helpful at all

**Comment by msc-** [comment]
The problem is possibly a bug in GYP. The libgl target is not yet in the list of targets GYP is building at the time target_conditions is evaluated. Weird as that is supposed to be evaluated "late".

I have found that using a regular `conditions` works but you have to move the `includes` out and use `library` instead of `_type`. See the new `issue-18` branch for sample code. The variable `library` is set in `gyp_includes/config.gypi`. It is `shared_library` for linux and OS X. You will probably need to copy the .so to the directory where the gl3loadtests executable is. There is a commented out copy block to point the way in lib/libktx.gypi.

**Comment by msc-** [comment]
I just pushed an update that makes the shared library fully functional on OS X.  You should be able to add the necessary flags and copy (if needed) for Linux fairly easily with this to follow.

**Comment by rjktcby** [comment]
I was able to build all with some changes and all tests passed.
You can check out  my diff in latest commit here: https://github.com/rjktcby/KTX/tree/issue-18-linux

I had to change `sdl_to_use` to `installed_dylib` for linux (and use libsdl from distro) as I didn't really get the point of these text files as libSDL.so. They don't link as they're not libraries. Is this some OSX magic?

**Comment by msc-** [comment]
Thank you for the changes. They look good with the exception of the change of `sdl_to_use` to `installed_dylib`. When you say "text files" I am not sure to what you are referring. When `built_dylib` is used, 2 files are copied `<(PRODUCT_DIR)`. One _is_ the library .so which has a full version number after the `.so`. The other is a symbolic link to the actual library using the name searched for by the linker. This is the standard `.so` naming scheme for Linux and it has been working fine for me on Ubuntu on which I have been building with both cmake and make and running the tests. It has nothing to do with OS X. What was the problem you were having?

I want to preserve `built_dylib` because I there are bug fixes in the built SDL included with KTX that enable me to reliably run tests using an OpenGL ES context as well as an OpenGL context. At present only `gl3loadtests` is being built for Linux but I am planning to at the `es[13]` tests.

**Comment by rjktcby** [comment]
When i say text files i mean that libSDL2-2.0.so.0.4.0 is a text file, other files are correct symlinks

    > cd other/lib/linux/Release-x64
    > file libSDL2-2.0.so
    libSDL2-2.0.so: symbolic link to libSDL2-2.0.so.0
    > file libSDL2-2.0.so.0
    libSDL2-2.0.so.0: symbolic link to libSDL2-2.0.so.0.4.0
    > file libSDL2-2.0.so.0.4.0 
    libSDL2-2.0.so.0.4.0: ASCII text
    > cat libSDL2-2.0.so.0.4.0 
    version https://git-lfs.github.com/spec/v1
    oid sha256:9a0a40d6f1f4f1c55c505f6bf563e4d435358d354cd083fd227c9369db1eb871
    size 6757110

So this .so cannot be linked and i get linker errors when buliding with `built_dylib`

I think it is supposed to be a real lib, according to BUILDING.md:

> Builds of SDL are provided in the KTX Git repo

**Comment by rjktcby** [comment]
Well i realized that i didn't install the git lfs extension, that's the problem, i'll check the builds with libs from repo and submit pull request. Should i do it against the issue-18 branch?

**Comment by msc-** [comment]
Yes it is Git LFS. I suddenly realized during the night and was about to add a comment to tell you. Please submit the pull request against the incoming branch.

**Comment by rjktcby** [comment]
Cmake build runs ok, but make build fails. It generates bad path to SDL libraries for the copy rules.
Right now i'm getting:

    $(builddir)/libSDL2-2.0.so.0.4.0: TOOLSET := $(TOOLSET)
    $(builddir)/libSDL2-2.0.so.0.4.0: other_lib/linux/$(BUILDTYPE)-x64/libSDL2-2.0.so.0.4.0 FORCE_DO_CMD
        $(call do_cmd,copy)

Which is obviously invalid build rule for building in `build/make/linux`
It should be either `../../../other_lib/linux/$(BUILDTYPE)-x64/libSDL2-2.0.so.0.4.0` or `$(srcdir)/other_lib/linux/$(BUILDTYPE)-x64/libSDL2-2.0.so.0.4.0`

Still not figured out how to chage gyp configs to make it work

**Comment by msc-** [comment]
You need to use my bug fixed [fork of GYP](https://github.com/msc-/gyp) that I pointed at in my first comment.

If you don't want to do that, you can submit a PR against the issue-18 branch instead and I will regenerate the make files and merge to incoming.

**Comment by rjktcby** [comment]
No, i'm fine with your fork, everything builds and all tests pass now

**Comment by msc-** [comment]
I have merged your pull request #19. Thanks.
